doc/protocol.md

```diff
@@ -829,6 +829,119 @@ or pipe is done.
+the case of directories, information can also be reported about the (directly)
+contained entries.
+
+Additional "open" command options should be specified with a channel of
+this payload type:
+
+ * `path` (mandatory): a string
+ * `attrs` (mandatory): a list of strings, the list of attributes to be
+   reported for the requested file.  See below for the possibilities.
+ * `fnmatch`: (optional, default ''): if non-empty, this is a [`fnmatch`-style
+   pattern](https://docs.python.org/3/library/fnmatch.html) selecting the directory
+   entries to be reported.  Use `*` to report all files.  Can also be used to
+   implement search (`searc*`) or hide temporary files (`[!.]*`).
+ * `close-on-error` (optional, default: true): if the channel should close
+   (with `problem`) on failure to `open()` the named `path`
+ * `watch` (optional, default: false): if the channel should watch for changes
+ * `follow` (default: true): if the trailing component of `path` is a symbolic
+   link, whether we should follow it.  `no-follow` mode is not supported for
+   `watch` (but is also mostly unnecessary).
+
+A channel is created with an associated absolute pathname ('path').  A path is
+a string that you pass to the `open()` syscall.  The object returned by that
+call is described by the (JSON Document) value of the channel.
+
+Strictly speaking, the path is specified to the channel as a unicode string,
+but paths on the filesystem are binary strings.  This mapping is handled using
+the usual Python convention: filenames are stored in utf8 and encoded and
+decoded with `errors='surrogateescape'`, which allows for handling of filenames
+which are not utf8.
+
+Conceptually, the channel represents the value of a JSON object describing what
+is found at the path, or the reason for an error.  Each message on the channel
+is a JSON Merge Patch (RFC7396) which updates or replaces this object.
+
+The top-level keys in the object are:
+
+ - `info`: the info about the reported path (described below)
+ - `problem`: a "problem code" like `not-found` or `not-directory`
+ - `message`: a user-readable (`strerror`) string describing the error
+ - `partial`: `true` if the data is in a "partial" state.  This is used when
+   sending updates in multiple pieces.
+
+If `close-on-error` is set, `problem` and `message` will never be sent, but the
+channel will be closed with the `problem` code set to indicate the error.
+
+Otherwise, the value of the `info` attribute is a JSON object describing the
+attributes of the file.  Any attribute which cannot be read (eg: due to
+permissions problems) is simply not reported.
+
+ * `type`: a string: `file`, `directory`, `link`, or `special`
+ * `mode`: an integer representing the "file permission" bits.  As per JSON,
+   this is transmitted as a decimal value, but it is meant to be interpreted in
+   octal, in the usual way.
+ * `uid`: an integer, the uid of the file owner (`st_uid`)
+ * `owner`: a string, or an integer if the lookup failed
```

Comment:
That's strange; are user names forbidden in principle to have only numeric characters? it's also inconsistent with "Any attribute which cannot be read is simply not reported". I'd rather leave it out, which makes the API more robust.

Reply:
If it reports the username then it's a string.  If it reports the uid, it's an integer, which is done in order to make the two cases easily distinguishable.

You can get the raw uid by requesting `uid`.

The idea here is that a hypothetical UI will probably want to display the username if it's available, and fall back to displaying the numeric UID otherwise.  If we don't implement that fallback here, then the user needs to subscribe to both `uid` and `user` in order to do that, which is 99% of the time redundant.

We could discuss always returning a string here, and if that string is numeric then the caller knows that they got a UID (and that's probably what they wanted to directly display anyway).  I personally prefer it as an int.

Reply:
I'm just concerned as this is JavaScript, which doesn't tend to make too fine a difference between them. `Number.isInteger()` does the right thing. OTOH `<span>{ props.owner ?? props.uid }</span>` feels more explicit to me, and makes it easier for UIs to actually tell whether a UID is valid or not (they may be formatted differently, for example).

Subscribing to both uid and user seems okay to me, if a program actually needs/wants that (most won't -- navigator is of course a case where it matters). But there is no additional overhead to detecting the uid.

Strict no for "always return a string". If you really like the int/string dual type here, then I don't veto it, but I prefer my types clean. :grin:

Reply:
Let's split the baby and introduce a third property!

 - `uid`: int
 - `user`: string or null
 - `uhh`: string or int

Reply:
Corresponding attribute for `gid` being called `guhh`, obviously.

Reply:
I guess my final word on this is: I don't know any situation where it would be useful to request `user` and *not* get the uid as a fallback in case the lookup fails, so I'm going to keep it the way it is.

Reply:
OK -- IMHO it makes the API harder and more brittle to use on the caller side, but at least it's possible.

Reply:
... in the sense of "requesting every single of those in `attrs` feels like overkill" -- you'd either request and get all ownership and permissions, or none of them, as they all come from a single stat() call anyway. I totally understand only getting dir contents and file contents upon request, but everything that comes from the same source should be treated as a property group.

Reply:
OK -- I can't say that a multi-typed value sparks joy in me from the API user POV, but it's good enough and it's your PR. I don't want to risk getting into baby splitting! :grin: